### PR TITLE
replace recaptcha with self-hosted captcha

### PR DESCRIPTION
**Pull request by chappjc**
This replaces Google's reCAPTCHA with a self hosted captcha.

**NOTE**:  This is needs review and testing.  It works for me though.

Create controllers/captcha.go, where the captcha GET and POST handlers
are implemented for retrieving images and verifying responses.
Create `CaptchaHandler` type with image size spec, and add it as a field
of `MainController`.
Create `ApplyCaptcha` captcha middleware.
Add GET route for `/captchas` for serving the captcha images.
Add POST route for `/verifyhuman` for verifying captcha responses and
redirecting to the previous page with the session updated.
Modify http handlers to display new captcha: `PasswordReset`, `SignUp`, and
`Settings`.
Modify http handlers to check for captcha verification:
`PasswordResetPost`, `SignUpPost`, and `SettingsPost`.
Unrelated: remove unused empty `responseHeaderMap`.
Comment on how insane `core/(*Application).Route` is.

### CODE REVIEW

**Comment by chappjc** [comment]
Force pushed https://github.com/decred/dcrstakepool/commit/996f7bf0ffc72354b2da0771564c58c86a4aef48.  Tested and working.  The pages don't look very pretty, but's that's about the best I can do.

File: controllers/captcha.go [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+	}
+
+	session := controller.GetSession(c)
+	w.Header().Set("Content-Type", "text/html; charset=utf-8")
+	var status int
+	if captcha.VerifyString(id, solution) {
+		session.Values["CaptchaDone"] = true
+		status = http.StatusFound
+	} else {
+		session.Values["CaptchaDone"] = false
```
Comment (chappjc) [inline controllers/captcha.go]: Should probably do a `session.AddFlash("Captcha verification failed. Try again.")` here too

File: controllers/captcha.go [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+	"net/http"
+	"path"
+	"strings"
+	"time"
+
+	"github.com/dchest/captcha"
+	"github.com/zenazn/goji/web"
+)
+
+type CaptchaHandler struct {
```
Comment (chappjc) [inline controllers/captcha.go]: I need to doc all these exported things.  And copyright the file.

File: controllers/main.go [hunk 3]
```diff
@@ -5,6 +5,7 @@ import (
 	"encoding/hex"
 	"errors"
 	"fmt"
+	"github.com/dchest/captcha"
```
Comment (chappjc) [inline controllers/main.go]: This import should be with the other github imports, not with the std lib imports.

File: controllers/main.go [hunk 4]
```diff
@@ -145,6 +143,11 @@ func NewMainController(params *chaincfg.Params, adminIPs []string,
 		return nil, err
 	}
 
+	ch := &CaptchaHandler{
+		ImgHeight: 127,
+		ImgWidth:  257,
```
Comment (chappjc) [inline controllers/main.go]: Pulled out of my ass and could be set in the config file, but that's not really needed now.

File: controllers/main.go [hunk 5]
```diff
@@ -1358,20 +1360,20 @@ func (controller *MainController) Index(c web.C, r *http.Request) (string, int)
...
-	t := controller.GetTemplate(c)
 	session := controller.GetSession(c)
 	c.Env["FlashError"] = session.Flashes("passwordresetError")
 	c.Env["FlashSuccess"] = session.Flashes("passwordresetSuccess")
 	c.Env["IsPasswordReset"] = true
-	c.Env["RecaptchaSiteKey"] = controller.recaptchaSiteKey
 	if controller.smtpHost == "" {
 		c.Env["SMTPDisabled"] = true
 	}
+	c.Env["CaptchaID"] = captcha.New()
```
Comment (chappjc) [inline controllers/main.go]: No reload button next to the captcha.  To get a new one, all you have to do is reload the page with the protected forms.

File: server.go [hunk 6]
```diff
@@ -227,6 +227,10 @@ func runMain() int {
 	app.Get("/signup", application.Route(controller, "SignUp"))
 	app.Post("/signup", application.Route(controller, "SignUpPost"))
 
+	// Captcha
+	app.Get("/captchas/*", controller.CaptchaServe)
+	app.Post("/verifyhuman", controller.CaptchaVerify)
```
Comment (chappjc) [inline server.go]: The `application.Route` function is a shit show.  Not using it.

File: system/core.go [hunk 7]
```diff
@@ -122,27 +122,21 @@ func (application *Application) Close() {
...
 		method := methodInterface.(func(c web.C, r *http.Request) (string, int))
 
 		body, code := method(c, r)
 
+		// Save the session in c.Env["Session"].
 		if err := saveSession(c, w, r); err != nil {
 			log.Errorf("Can't save session: %v", err)
 		}
 
-		if respHeader, exists := c.Env["ResponseHeaderMap"]; exists {
```
Comment (chappjc) [inline system/core.go]: This was a relic from a much older feature that was partially ripped out.

**Comment by honeycrypto** [comment]
Why to overcomplicate things? Recaptcha works great.

**Comment by chappjc** [comment]
@honeycrypto Please see https://github.com/decred/dcrstakepool/issues/279

**Comment by teknico** [comment]
This branch works for me too.

Under a local deployment I was able to sign up with the captcha image: I got the confirmation email and activated the account. Looking at the database I confirmed that the account got id no. 1 (configured as admin), and logging in I was indeed able to see the low-fee ticket page.

(The local deployment consists of one dcrd instance, one dcrwallet for fees, two voting dcrwallet, two stakepoold, one dcrstakepool and one mysqld. To make all the daemons run in the same namespace without virtual machines nor containers, I changed the various listening ports in the daemons' configuration.)

**Comment by chappjc** [comment]
Thanks for the feedback @teknico.

**Comment by dajohi** [comment]
@chappjc please rebase on top of current master

**Comment by chappjc** [comment]
@dajohi rebased

**Comment by chappjc** [comment]
@dajohi rebased again

**Pull request merged**